packages/sources/finalto/src/transport/forex.ts

```diff
@@ -0,0 +1,161 @@
+  const day = dateLike.substring(6, 8)
+  const rest = dateLike.slice(8)
+  return new Date(`${year}.${month}.${day}${rest}Z`).getTime()
+}
+
+export const wsTransport: WebsocketReverseMappingTransport<WsTransportTypes, string> =
+  new WebsocketReverseMappingTransport<WsTransportTypes, string>({
+    url: (context) => context.adapterSettings.WS_API_ENDPOINT,
+    handlers: {
+      open(connection, context) {
+        return new Promise((resolve) => {
+          connection.addEventListener('message', (event: MessageEvent) => {
+            const parsed = JSON.parse(event.data.toString())
+            if (parsed.MsgType === 'Logon') {
+              logger.info('Got logged in response, connection is ready')
+              resolve()
+            }
+          })
+          connection.send(
+            JSON.stringify({
+              MsgType: 'Logon',
+              Username: context.adapterSettings.WS_API_USERNAME,
+              Password: context.adapterSettings.WS_API_PASSWORD,
+            }),
+          )
+        })
+      },
+      message(message) {
+        if (message.MsgType !== 'SymbolPrices') {
+          logger.warn(JSON.stringify(message), 'Unexpected message')
+          return
+        }
+
+        const pair = wsTransport.getReverseMapping(message.Symbol)
+
+        if (!pair) {
+          logger.warn(`Reverse mapping pair for '${message.Symbol}' was not found`)
+          return
+        }
+
+        if (!message.Prices.length) {
+          return [
+            {
+              params: { base: pair.base, quote: pair.quote },
+              response: {
+                errorMessage: `Missing price info for '${pair.base}${pair.quote}'`,
+                statusCode: 502,
+              },
+            },
+          ]
+        }
+
+        const bidInfo = message.Prices.find((p) => p.Type === 'B')
+        const askInfo = message.Prices.find((p) => p.Type === 'O')
+
+        if (!bidInfo || !askInfo || !bidInfo.Px || !askInfo.Px) {
+          return [
+            {
+              params: { base: pair.base, quote: pair.quote },
+              response: {
+                errorMessage: `Missing bid and/or ask prices for '${pair.base}${pair.quote}'`,
+                statusCode: 502,
+              },
+            },
+          ]
+        }
+
+        const result = (Number(bidInfo.Px) + Number(askInfo.Px)) / 2
+
+        return [
+          {
+            params: { base: pair.base, quote: pair.quote },
+            response: {
+              result,
+              data: {
+                result,
+              },
+              timestamps: {
+                providerIndicatedTimeUnixMs: parseDate(message.SendTime),
+              },
+            },
+          },
+        ]
+      },
+    },
+    builders: {
+      subscribeMessage: (params) => {
+        const symbol = `${params.base}${params.quote}`.toUpperCase()
+        wsTransport.setReverseMapping(symbol, params)
+        const id = getSubscriptionId(symbol)
```

Comment:
What happens if multiple EA's use the same API key, and attempt to use the same subscription ID? 

Could we instead make this a UUID rather than an incrementing integer? If not, some explicit error handling for the conflicting subscription ID case would be useful.

Reply:
@alecgard AFAIK each set of credentials have a unique WS address that allows only one connection, but i think you are right, that UUID might be better, as we won't think about collision.

Reply:
That makes sense, just seeing your comment about having a subscription id of `1` causing errors made me suspicious that the same credential may be being used in multiple places